File: test/rb/t-readstring.rb
require_relative 'utils'

t = DbgScript.current_thread
f = t.current_frame
locals = f.get_locals

# Get the 'car' local.
#
car = locals.find {|t| t.name == 'car'}

# Read ANSI string.
#
puts car['name'].read_string      # until NUL
puts car['name'].read_string(1)   # first char
puts car['name'].read_string(2)   # first two chars
puts car['name'].read_string(-1)  # until NUL

# Longer than the real length. Expect full string to be read without any embedded
# NULs.
#
puts car['name'].read_string(1000)

# Exercise DbgScript#read_string.
#
puts DbgScript.read_string(car['name'].address)
puts DbgScript.read_string(car['name'].address, 1)
puts DbgScript.read_string(car['name'].address, 5)
puts DbgScript.read_string(car['name'].address, 500)
puts DbgScript.read_string(car['name'].address, -1)

# Negative tests.
#
negative_test(ArgumentError) {
  puts car['name'].read_string(100000000)  # too big
}
negative_test(ArgumentError) {
  puts car['name'].read_string(0)  # 0 invalid
}
negative_test(ArgumentError) {
  puts DbgScript.read_string(car['name'].address, 0)
}
negative_test(ArgumentError) {
  puts DbgScript.read_string(car['name'].address, 10000000)
}

# Read Wide string.
#
puts car['wide_name'].read_wide_string      # until NUL
puts car['wide_name'].read_wide_string(1)   # first char
puts car['wide_name'].read_wide_string(2)   # first two chars
puts car['wide_name'].read_wide_string(-1)  # until NUL

# Longer than the real length. Expect full string to be read without any embedded
# NULs.
#
puts car['wide_name'].read_wide_string(1000)

# Negative test.
#
negative_test(ArgumentError) {
  puts car['wide_name'].read_wide_string(0)
}
negative_test(ArgumentError) {
  puts car['wide_name'].read_wide_string(1000000)
}

# Exercise DbgScript#read_wide_string.
#
puts DbgScript.read_wide_string(car['wide_name'].address)
puts DbgScript.read_wide_string(car['wide_name'].address, 1)
puts DbgScript.read_wide_string(car['wide_name'].address, 5)
puts DbgScript.read_wide_string(car['wide_name'].address, 500)
puts DbgScript.read_wide_string(car['wide_name'].address, -1)

File: test/lua/t-readstring.lua
require 'utils'

local car = getCar()

-- Read ANSI string.
--
print(car:f('name'):readString())      -- until NUL
print(car:f('name'):readString(1))     -- first char
print(car:f('name'):readString(2))     -- first two chars
print(car:f('name'):readString(-1))    -- until NUL

-- Longer than the real length. Expect full string to be read without any
-- embedded NULs.
--
print(car:f('name'):readString(1000))

-- Exercise dbgscript.readString.
--
print(dbgscript.readString(car:f('name').address))
print(dbgscript.readString(car:f('name').address, 1))
print(dbgscript.readString(car:f('name').address, 5))
print(dbgscript.readString(car:f('name').address, 500))
print(dbgscript.readString(car:f('name').address, -1))

-- Read Wide string.
--
print(car:f('wide_name'):readWideString())      -- until NUL
print(car:f('wide_name'):readWideString(1))    -- first char
print(car:f('wide_name'):readWideString(2))    -- first two chars
print(car:f('wide_name'):readWideString(-1))   -- until NUL

-- Longer than the real length. Expect full string to be read without any embedded
-- NULs.
--
print(car:f('wide_name'):readWideString(1000))

-- Exercise dbgscript.readWideWideString.
--
print(dbgscript.readWideString(car:f('wide_name').address))
print(dbgscript.readWideString(car:f('wide_name').address, 1))
print(dbgscript.readWideString(car:f('wide_name').address, 5))
print(dbgscript.readWideString(car:f('wide_name').address, 500))
print(dbgscript.readWideString(car:f('wide_name').address, -1))
